flocker/node/agents/functional/test_ebs.py

```diff
@@ -156,3 +162,293 @@ def test_next_device_in_use(self):
+                                state_flow.end_state])
+
+            err_states = set(VolumeStates._enumerants.values()) - valid_states
+            err_state = err_states.pop()
+            return err_state.value
+        elif state_type == self.S.IN_TRANSIT:
+            return state_flow.transient_state.value
+        elif state_type == self.S.DESTINATION:
+            return state_flow.end_state.value
+
+    def _pick_attach_data(self, attach_type):
+        """
+        Helper function to create desired volume attach data.
+
+        :param NamedConstant attach_type: Type of attach data to create.
+
+        :returns: Volume attachment set that conforms to requested attach type.
+        :rtype: AttachmentSet
+        """
+        if attach_type == self.A.MISSING_ATTACH_DATA:
+            return None
+        elif attach_type == self.A.MISSING_INSTANCE_ID:
+            attach_data = AttachmentSet()
+            attach_data.device = u'/dev/sdf'
+            attach_data.instance_id = ''
+            return attach_data
+        elif attach_type == self.A.MISSING_DEVICE:
+            attach_data = AttachmentSet()
+            attach_data.device = ''
+            attach_data.instance_id = u'i-xyz'
+            return attach_data
+        elif attach_type == self.A.ATTACH_SUCCESS:
+            attach_data = AttachmentSet()
+            attach_data.device = u'/dev/sdf'
+            attach_data.instance_id = u'i-xyz'
+            return attach_data
+        elif attach_type == self.A.DETACH_SUCCESS:
+            return None
+
+    def _custom_update(self, operation, state_type,
+                       attach_data=A.MISSING_ATTACH_DATA):
+        """
+        Create a custom update function for a volume.
+        """
+        def update(volume):
+            """
+            Transition volume to desired end state and attach data.
+
+            :param boto.ec2.volume.Volume volume: Volume to move to
+                invalid state.
+            """
+            volume.status = self._pick_end_state(operation, state_type)
+            volume.attach_data = self._pick_attach_data(attach_data)
+        return update
+
+    def _assert_fail(self, operation, testcase,
+                     attach_data_type=A.MISSING_ATTACH_DATA):
+        """
+        Assert that configured volume state change for given testcase indicates
+        incomplete operation execution.
+        """
+        volume = self._create_template_ebs_volume(operation)
+        update = self._custom_update(operation, testcase, attach_data_type)
+        start_time = time.time()
+
+        if testcase == self.S.ERROR:
+            self.assertRaises(UnexpectedStateException, _should_finish,
+                              operation, volume, update, start_time, TIMEOUT)
+        else:
+            finish_result = _should_finish(operation, volume, update,
+                                           start_time)
+            self.assertEqual(False, finish_result)
```

Comment:
Apologies for poor naming. ``testcase`` was a flag to indicate type of volume state transition test. Renamed ``testcase``, and separated out ``ERROR_STATE`` assert from assert module for other volume end states (``TRANSIT_STATE``, ``DESTINATION_STATE``).